Add expression index in multi-schema change panic

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
drop table if exists t;
create table t (a int, b int);
insert into t values (1, 2), (2, 1);
alter table t add column c int default 10, add index idx1((a + b)), add unique index idx2((a + b));
```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

```
ERROR 1062 (23000): Duplicate entry '3' for key 't.idx2'
```

### 3. What did you see instead (Required)

```
ERROR 1105 (HY000): When the state is JobStateRollbackDone or JobStateCancelled, historyJob.Error should never be nil
```

Log output
```
[2022/06/15 10:27:01.531 +08:00] [ERROR] [misc.go:109] ["panic in the recoverable goroutine"] [label=ddl-worker] [funcInfo="worker 2, tp reorg runDDLJob"] [r="\"invalid memory address or nil pointer dereference\""] [stack="github.com/pingcap/tidb/util.Recover
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/util/misc.go:113
runtime.gopanic
  /opt/homebrew/Cellar/go/1.18.1/libexec/src/runtime/panic.go:844
runtime.panicmem
  /opt/homebrew/Cellar/go/1.18.1/libexec/src/runtime/panic.go:220
runtime.sigpanic
  /opt/homebrew/Cellar/go/1.18.1/libexec/src/runtime/signal_unix.go:818
github.com/pingcap/tidb/parser/model.FindColumnInfo
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/parser/model/model.go:285
github.com/pingcap/tidb/ddl.getPrimaryKey
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/ddl/ddl_api.go:1662
github.com/pingcap/tidb/ddl.checkInvisibleIndexOnPK
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/ddl/ddl_api.go:1631
github.com/pingcap/tidb/ddl.(*worker).onCreateIndex
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/ddl/index.go:542
github.com/pingcap/tidb/ddl.(*worker).runDDLJob
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/ddl/ddl_worker.go:893
github.com/pingcap/tidb/ddl.onMultiSchemaChange
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/ddl/multi_schema_change.go:94
github.com/pingcap/tidb/ddl.(*worker).runDDLJob
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/ddl/ddl_worker.go:959
github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue.func1
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/ddl/ddl_worker.go:645
github.com/pingcap/tidb/kv.RunInNewTxn
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/kv/txn.go:127
github.com/pingcap/tidb/ddl.(*worker).handleDDLJobQueue
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/ddl/ddl_worker.go:598
github.com/pingcap/tidb/ddl.(*worker).start
  /Users/tangenta/gopath/src/github.com/pingcap/tidb/ddl/ddl_worker.go:211"]

```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
a75690f17

--- a/ddl/index.go
+++ b/ddl/index.go
@@ -419,25 +419,30 @@ func checkPrimaryKeyNotNull(d *ddlCtx, w *worker, sqlMode mysql.SQLMode, t *meta
 	return nil, err
 }
 
-func updateHiddenColumns(tblInfo *model.TableInfo, idxInfo *model.IndexInfo, state model.SchemaState) *model.IndexInfo {
-	// Find the last public column.
-	pos := 0
+func updateHiddenColumnsToPublic(tblInfo *model.TableInfo, idxInfo *model.IndexInfo) {
+	hiddenColOffset := make(map[int]struct{}, 0)
+	for _, col := range idxInfo.Columns {
+		if tblInfo.Columns[col.Offset].Hidden {
+			hiddenColOffset[col.Offset] = struct{}{}
+		}
+	}
+	if len(hiddenColOffset) == 0 {
+		return
+	}
+	// Find the first non-public column.
+	firstNonPublicPos := len(tblInfo.Columns) - 1
 	for i, c := range tblInfo.Columns {
-		if c.State == model.StatePublic {
-			pos = i
-		} else {
+		if c.State != model.StatePublic {
+			firstNonPublicPos = i
 			break
 		}
 	}
 	for _, col := range idxInfo.Columns {
-		if tblInfo.Columns[col.Offset].Hidden {
-			tblInfo.Columns[col.Offset].State = state
+		tblInfo.Columns[col.Offset].State = model.StatePublic
+		if _, needMove := hiddenColOffset[col.Offset]; needMove {
+			tblInfo.MoveColumnInfo(col.Offset, firstNonPublicPos)
 		}
 	}
-	for _, col := range idxInfo.Columns {
-		tblInfo.MoveColumnInfo(col.Offset, pos)
-	}
-	return tblInfo.FindIndexByName(idxInfo.Name.L)
 }
 
 func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK bool) (ver int64, err error) {
@@ -563,7 +568,7 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
 	case model.StateNone:
 		// none -> delete only
 		indexInfo.State = model.StateDeleteOnly
-		indexInfo = updateHiddenColumns(tblInfo, indexInfo, model.StatePublic)
+		updateHiddenColumnsToPublic(tblInfo, indexInfo)
 		ver, err = updateVersionAndTableInfoWithCheck(d, t, job, tblInfo, originalState != indexInfo.State)
 		if err != nil {
 			return ver, err

--- a/ddl/index.go
+++ b/ddl/index.go
@@ -419,12 +419,25 @@ func checkPrimaryKeyNotNull(d *ddlCtx, w *worker, sqlMode mysql.SQLMode, t *meta
 	return nil, err
 }
 
-func updateHiddenColumns(tblInfo *model.TableInfo, idxInfo *model.IndexInfo, state model.SchemaState) {
+func updateHiddenColumns(tblInfo *model.TableInfo, idxInfo *model.IndexInfo, state model.SchemaState) *model.IndexInfo {
+	// Find the last public column.
+	pos := 0
+	for i, c := range tblInfo.Columns {
+		if c.State == model.StatePublic {
+			pos = i
+		} else {
+			break
+		}
+	}
 	for _, col := range idxInfo.Columns {
 		if tblInfo.Columns[col.Offset].Hidden {
 			tblInfo.Columns[col.Offset].State = state
 		}
 	}
+	for _, col := range idxInfo.Columns {
+		tblInfo.MoveColumnInfo(col.Offset, pos)
+	}
+	return tblInfo.FindIndexByName(idxInfo.Name.L)
 }
 
 func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK bool) (ver int64, err error) {
@@ -550,7 +563,7 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
 	case model.StateNone:
 		// none -> delete only
 		indexInfo.State = model.StateDeleteOnly
-		updateHiddenColumns(tblInfo, indexInfo, model.StatePublic)
+		indexInfo = updateHiddenColumns(tblInfo, indexInfo, model.StatePublic)
 		ver, err = updateVersionAndTableInfoWithCheck(d, t, job, tblInfo, originalState != indexInfo.State)
 		if err != nil {
 			return ver, err

--- a/ddl/multi_schema_change_test.go
+++ b/ddl/multi_schema_change_test.go
@@ -1079,7 +1079,7 @@ func TestMultiSchemaChangeAlterIndexVisibility(t *testing.T) {
 }
 
 func TestMultiSchemaChangeWithExpressionIndex(t *testing.T) {
-	store, clean := testkit.CreateMockStore(t)
+	store, dom, clean := testkit.CreateMockStoreAndDomain(t)
 	defer clean()
 	tk := testkit.NewTestKit(t, store)
 	tk.MustExec("use test;")
@@ -1087,7 +1087,36 @@ func TestMultiSchemaChangeWithExpressionIndex(t *testing.T) {
 	tk.MustExec("create table t (a int, b int);")
 	tk.MustExec("insert into t values (1, 2), (2, 1);")
 	tk.MustGetErrCode("alter table t drop column a, add unique index idx((a + b));", errno.ErrUnsupportedDDLOperation)
-	tk.MustGetErrCode("alter table t add column c int, change column a d bigint, add index idx((a + a))", errno.ErrUnsupportedDDLOperation)
+	tk.MustGetErrCode("alter table t add column c int, change column a d bigint, add index idx((a + a));", errno.ErrUnsupportedDDLOperation)
+	tk.MustGetErrCode("alter table t add column c int default 10, add index idx1((a + b)), add unique index idx2((a + b));",
+		errno.ErrDupEntry)
+	tk.MustQuery("select * from t;").Check(testkit.Rows("1 2", "2 1"))
+
+	originHook := dom.DDL().GetHook()
+	hook := &ddl.TestDDLCallback{Do: dom}
+	var checkErr error
+	hook.OnJobRunBeforeExported = func(job *model.Job) {
+		assert.Equal(t, model.ActionMultiSchemaChange, job.Type)
+		if job.MultiSchemaInfo.SubJobs[1].SchemaState == model.StateWriteOnly {
+			tk2 := testkit.NewTestKit(t, store)
+			tk2.MustExec("use test;")
+			_, checkErr = tk2.Exec("update t set a = 3 where a = 1;")
+			if checkErr != nil {
+				return
+			}
+			_, checkErr = tk2.Exec("insert into t values (10, 10);")
+		}
+	}
+	dom.DDL().SetHook(hook)
+	tk.MustExec("alter table t add column c int default 10, add index idx1((a + b)), add unique index idx2((a + b));")
+	require.NoError(t, checkErr)
+	dom.DDL().SetHook(originHook)
+
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (a int, b int);")
+	tk.MustExec("insert into t values (1, 2), (2, 1);")
+	tk.MustExec("alter table t add column c int default 10, add index idx1((a + b)), add unique index idx2((a*10 + b));")
+	tk.MustQuery("select * from t use index(idx1, idx2);").Check(testkit.Rows("1 2 10", "2 1 10"))
 }
 
 func composeHooks(dom *domain.Domain, cbs ...ddl.Callback) ddl.Callback {

--- a/planner/core/logical_plan_builder.go
+++ b/planner/core/logical_plan_builder.go
@@ -5277,7 +5277,7 @@ func (b *PlanBuilder) buildUpdateLists(ctx context.Context, tableList []*ast.Tab
 		if !found {
 			return nil, nil, false, infoschema.ErrTableNotExists.GenWithStackByArgs(tn.DBInfo.Name.O, tableInfo.Name.O)
 		}
-		for i, colInfo := range tableInfo.Columns {
+		for i, colInfo := range tableVal.Cols() {
 			if !colInfo.IsGenerated() {
 				continue
 			}

